cmake/kokkos_functions.cmake

```diff
@@ -379,70 +547,73 @@ MACRO(kokkos_find_imported NAME)
 
   SET(${NAME}_INCLUDE_DIRS)
   IF (TPL_HEADER)
-    IF(TPL_HEADER_PATHS)
-      FIND_PATH(${NAME}_INCLUDE_DIRS ${TPL_HEADER} PATHS ${TPL_HEADER_PATHS})
-    ELSE()
-      FIND_PATH(${NAME}_INCLUDE_DIRS ${TPL_HEADER} PATHS ${${NAME}_ROOT}/include ${KOKKOS_${NAME}_DIR}/include)
-    ENDIF()
+    KOKKOS_FIND_HEADER(${NAME}_INCLUDE_DIRS ${TPL_HEADER} ${NAME} ${ALLOW_PATH_FALLBACK_OPT} PATHS ${TPL_HEADER_PATHS})
   ENDIF()
 
   FOREACH(HEADER ${TPL_HEADERS})
-    IF(TPL_HEADER_PATHS)
-      FIND_LIBRARY(HEADER_FIND_TEMP ${HEADER} PATHS ${TPL_HEADER_PATHS})
-    ELSE()
-      FIND_LIBRARY(HEADER_FIND_TEMP ${HEADER} PATHS ${${NAME}_ROOT}/lib ${KOKKOS_${NAME}_DIR}/lib)
-    ENDIF()
+    KOKKOS_FIND_HEADER(HEADER_FIND_TEMP ${HEADER} ${NAME} ${ALLOW_PATH_FALLBACK_OPT} PATHS ${TPL_HEADER_PATHS})
     IF(HEADER_FIND_TEMP)
       LIST(APPEND ${NAME}_INCLUDE_DIRS ${HEADER_FIND_TEMP})
     ENDIF()
   ENDFOREACH()
 
   SET(${NAME}_LIBRARY)
   IF(TPL_LIBRARY)
-    IF(TPL_LIBRARY_PATHS)
-      FIND_LIBRARY(${NAME}_LIBRARY ${TPL_LIBRARY} PATHS ${TPL_LIBRARY_PATHS})
-    ELSE()
-      FIND_LIBRARY(${NAME}_LIBRARY ${TPL_LIBRARY} PATHS ${${NAME}_ROOT}/lib ${KOKKOS_${NAME}_DIR}/lib)
-    ENDIF()
+    KOKKOS_FIND_LIBRARY(${NAME}_LIBRARY ${TPL_LIBRARY} ${NAME} ${ALLOW_PATH_FALLBACK_OPT} PATHS ${TPL_LIBRARY_PATHS})
   ENDIF()
 
-  SET(${NAME}_LIBRARIES)
+  SET(${NAME}_FOUND_LIBRARIES)
   FOREACH(LIB ${TPL_LIBRARIES})
-    IF(TPL_LIBRARY_PATHS)
-      FIND_LIBRARY(${LIB}_LOCATION ${LIB} PATHS ${TPL_LIBRARY_PATHS})
-    ELSE()
-      FIND_LIBRARY(${LIB}_LOCATION ${LIB} PATHS ${${NAME}_ROOT}/lib ${KOKKOS_${NAME}_DIR}/lib)
-    ENDIF()
+    KOKKOS_FIND_LIBRARY(${LIB}_LOCATION ${LIB} ${NAME} ${ALLOW_PATH_FALLBACK_OPT} PATHS ${TPL_LIBRARY_PATHS})
     IF(${LIB}_LOCATION)
-      LIST(APPEND ${NAME}_LIBRARIES ${${LIB}_LOCATION})
+      LIST(APPEND ${NAME}_FOUND_LIBRARIES ${${LIB}_LOCATION})
     ELSE()
-      SET(${NAME}_LIBRARIES ${${LIB}_LOCATION}) 
+      SET(${NAME}_FOUND_LIBRARIES ${${LIB}_LOCATION}) 
       BREAK()
     ENDIF()
   ENDFOREACH()
 
   INCLUDE(FindPackageHandleStandardArgs)
+  # I don't know a good way to chain all of the relevant variables into
+  # a single call to find_package_handle. find_package only cares
+  # about the failure of the LAST call to find_package_handle_...
+  # so if there as an error I need to skip the rest
```

Comment:
Look at the 2nd signature in the [doc](https://cmake.org/cmake/help/latest/module/FindPackageHandleStandardArgs.html?highlight=find_package_handle_standard_args#command:find_package_handle_standard_args), what you want is `REQUIRED_VARS <required-var>...`

Reply:
Fixed. I love CMake argument packing.